src/bento/components/bento-autocomplete/1.0/component.tsx

```diff
@@ -0,0 +1,555 @@
+  useEffect,
+  useMemo,
+  useRef,
+  useState,
+} from '#preact';
+import {ContainWrapper, useValueRef} from '#preact/component';
+import {useQuery} from '#preact/hooks/useQuery';
+import {xhrUtils} from '#preact/utils/xhr';
+
+import fuzzysearch from '#third_party/fuzzysearch';
+
+import {addParamToUrl} from 'src/url';
+
+import {useStyles} from './component.jss';
+import {
+  getEnabledResults,
+  getItemElement,
+  getSelectedObjectValue,
+  getSelectedTextValue,
+} from './helpers';
+import {tokenPrefixMatch} from './token-prefix-match';
+import {
+  BentoAutocompleteProps,
+  InputElement,
+  Item,
+  ItemTemplateProps,
+  OnSelectData,
+  isValidFilterType,
+} from './types';
+import {useAutocompleteBinding} from './use-autocomplete-binding';
+
+const INITIAL_ACTIVE_INDEX = -1;
+
+const getItems = (response: {items: Item[]}) =>
+  getValueForExpr(response, 'items');
+
+/**
+ * @param {!BentoAutocomplete.Props} props
+ * @return {PreactDef.Renderable}
+ */
+export function BentoAutocomplete({
+  id,
+  children,
+  onError = () => {},
+  onSelect = () => {},
+  filter = 'none',
+  minChars = 1,
+  items = [],
+  fetchJson = xhrUtils.fetchJson,
+  filterValue = 'value',
+  maxItems,
+  highlightUserEntry = false,
+  inline,
+  itemTemplate,
+  parseJson = getItems,
+  suggestFirst = false,
+  src,
+  query,
+}: BentoAutocompleteProps) {
+  const elementRef = useRef<HTMLElement>(null);
+  const containerId = useValueRef<string>(
+    id || `${Math.floor(Math.random() * 100)}_AMP_content_`
+  );
+  const inputRef = useRef<HTMLInputElement>(null);
+  const containerRef = useRef<HTMLDivElement>(null);
+
+  const [substring, setSubstring] = useState<string>('');
+  const [activeIndex, setActiveIndex] = useState<number>(INITIAL_ACTIVE_INDEX);
+  const [areResultsDisplayed, setAreResultsDisplayed] =
+    useState<boolean>(false);
+  const [shouldSuggestFirst, setShouldSuggestFirst] =
+    useState<boolean>(suggestFirst);
+  const classes = useStyles();
+  const [hasFetchedInitialData, setHasFetchedInitialData] = useState(false);
+  // In order to enable refetching, we currently have to disable the request between
+  // and re-enable it on user input.
+  const [hasFetchedData, setHasFetchedData] = useState(false);
+
+  const binding = useAutocompleteBinding(inline);
+
+  const url = useMemo(() => {
+    if (src && query) {
+      // Don't create an invalid URL if there is a query key but not a value
+      return substring ? addParamToUrl(src, query, substring) : null;
+    }
+    return src;
+  }, [src, substring, query]);
+
+  const isFetchEnabled = useMemo(() => {
+    if (url) {
+      if (query) {
+        return !hasFetchedData;
+      }
+      return !hasFetchedInitialData;
+    }
+    return false;
+  }, [url, hasFetchedInitialData, hasFetchedData, query]);
+
+  const shouldRefetch = useMemo(() => {
+    return !!query;
+  }, [query]);
+
+  const {data} = useQuery<Item[]>(
+    async () => {
+      const response = await fetchJson(url!);
```

Comment:
I just recently discovered we can't use async/await in our components. We'll want to refactor this to use promises. I'm actually working right now on refactoring bento-list to use promises.